### PR TITLE
METRON-1419: Create a SolrDao

### PR DESCRIPTION
**Pull request by merrimanr**
## Contributor Comments
This PR is an initial attempt at creating a SolrDao that implements the IndexDao interface, is functionally equivalent to ElasticsearchDao and passes all tests in SearchIntegrationTest and UpdateIntegrationTest.  

A high level summary of the changes include:

- Upgraded the Solr client version to 6.6.0
- Updated the SolrComponent to work with Solr 6.6 and added a couple convenience methods similar to ElasticsearchComponent
- Added a new SolrDao implementation with all IndexDao methods implemented
- Refactored the SearchIntegrationTest to work for both Solr and Elasticsearch and added an Solr implementation (more detail below)
- Created an abstract UpdateIntegrationTest and added a Solr implementation
- Added Solr schemas for test data sets
- Added new tests to SearchIntegrationTest including filtering on fields with different types, faceting on fields with different types, and faceting on fields with missing types.
- Broke the IndexDao down in the SolrDao to smaller, easier to understand classes.  The ElasticsearchDao class has become very large so I attempted to make the SolrDao more readable.

There were a couple areas where Elasticsearch and Solr behave slightly different.  I attempted to accommodated for that through the SolrDao implementation, by adjusting existing tests, and by splitting out specific tests:

- Column metadata is different between the 2 search engines so each implementation has their own tests
- There are cases where the clients will return different types in search results.  I am handling this in SearchIntegrationTest by first converting the types to strings and then comparing (other ideas?).  For example, the ES client returns an Integer for timestamp while Solr returns a Long.
- There are cases where ES throws an error under certain conditions while Solr does not (and vice-versa).  These were moved to either ES or Solr SearchIntegrationTest implementations.
- There is no support in Solr for sorting group results so I am sorting them client-side instead.

At this point the scope is limited to tests passing, meaning metron-elasticsearch and metron-solr pass all tests.  There are other PRs in progress that are needed before automated testing with full dev can be done.  I am still actively working on manually testing in full dev and adding documentation but this should get us started.

This PR is intended to be merged into the upgrade-solr feature branch but I have it set to master temporarily so review is easier.  We will need to merge in master to the feature branch to get rid of the extra commits since this PR is up to date with master.

I'm expecting a lengthy review and would request multiple reviewers and +1s.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel).
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

File: metron-platform/metron-indexing/src/main/java/org/apache/metron/indexing/dao/search/SearchDao.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.metron.indexing.dao.search;
+
+import java.io.IOException;
+import java.util.List;
+import org.apache.metron.indexing.dao.update.Document;
+
+public interface SearchDao {
```
Comment (justinleet) [inline metron-platform/metron-indexing/src/main/java/org/apache/metron/indexing/dao/search/SearchDao.java]: Seems like if we're adding this (and I do like having this split out by functionality), it should also be applied to the Elasticsearch Daos (same for things like UpdateDao).

I'd also strongly be in favor of splitting out the ElasticsearchDao as part of this, to make sure that the overall structure of the two Dao sets are very comparable and easy to understand.  The integration tests should be able to insulate us from the obvious issues, and IMO it's worthwhile to do now.

File: metron-platform/metron-indexing/src/test/java/org/apache/metron/indexing/dao/SearchIntegrationTest.java [hunk 2]
```diff
@@ -443,11 +495,11 @@ public void all_query_returns_all_results() throws Exception {
     Assert.assertEquals(10, results.size());
     for(int i = 0;i < 5;++i) {
       Assert.assertEquals("snort", results.get(i).getSource().get("source:type"));
-      Assert.assertEquals(10 - i, results.get(i).getSource().get("timestamp"));
+      Assert.assertEquals(10 - i + "", results.get(i).getSource().get("timestamp").toString());
```
Comment (justinleet) [inline metron-platform/metron-indexing/src/test/java/org/apache/metron/indexing/dao/SearchIntegrationTest.java]: Why are all the timestamps strings now?

File: metron-platform/metron-solr/src/main/java/org/apache/metron/solr/dao/SolrSearchDao.java [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+      LOG.error(msg, e);
+      throw new InvalidSearchException(msg, e);
+    }
+  }
+
+  @Override
+  public GroupResponse group(GroupRequest groupRequest) throws InvalidSearchException {
+    String groupNames = groupRequest.getGroups().stream().map(Group::getField).collect(
+        Collectors.joining(","));
+    SolrQuery query = new SolrQuery()
```
Comment (justinleet) [inline metron-platform/metron-solr/src/main/java/org/apache/metron/solr/dao/SolrSearchDao.java]: The query stuff is hardcoded right now.  Should just need to grab from the groupRequest.

File: metron-platform/metron-indexing/src/main/java/org/apache/metron/indexing/dao/search/SearchDao.java (hunk 1 above)
Comment (merrimanr) [reply metron-platform/metron-indexing/src/main/java/org/apache/metron/indexing/dao/search/SearchDao.java]: I was undecided as to whether we should do it in this PR or a future one and also wanted to make sure people liked it.  Will make this change.

File: metron-platform/metron-indexing/src/test/java/org/apache/metron/indexing/dao/SearchIntegrationTest.java (hunk 2 above)
Comment (merrimanr) [reply metron-platform/metron-indexing/src/test/java/org/apache/metron/indexing/dao/SearchIntegrationTest.java]: The ES and Solr clients return different types (Int vs Longs) so I am converting to string to account for that.  I'm not sure this is the correct approach and open to ideas.

File: metron-platform/metron-indexing/src/test/java/org/apache/metron/indexing/dao/SearchIntegrationTest.java [hunk 4]
```diff
@@ -655,83 +699,54 @@ public void disabled_facet_query_returns_null_count() throws Exception {
...
+  public void missing_type_facet_query() throws Exception {
+    SearchRequest request = JSONUtils.INSTANCE.load(missingTypeFacetQuery, SearchRequest.class);
+    SearchResponse response = dao.search(request);
+    Assert.assertEquals(10, response.getTotal());
+
+    Map<String, Map<String, Long>> facetCounts = response.getFacetCounts();
+    Assert.assertEquals(1, facetCounts.size());
+    Map<String, Long> snortFieldCounts = facetCounts.get("snort_field");
+    Assert.assertEquals(5, snortFieldCounts.size());
+    Assert.assertEquals(new Long(1), snortFieldCounts.get("50"));
```
[the remark this answers is not in the document]
Comment (merrimanr) [reply metron-platform/metron-indexing/src/test/java/org/apache/metron/indexing/dao/SearchIntegrationTest.java]: Done

File: metron-platform/metron-solr/src/main/java/org/apache/metron/solr/dao/SolrDao.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+import org.apache.metron.indexing.dao.update.Document;
+import org.apache.solr.client.solrj.SolrClient;
+import org.apache.solr.client.solrj.impl.CloudSolrClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SolrDao implements IndexDao {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  public static final String ID_FIELD = "guid";
```
[the remark this answers is not in the document]
Comment (merrimanr) [reply metron-platform/metron-solr/src/main/java/org/apache/metron/solr/dao/SolrDao.java]: Done

File: metron-platform/metron-solr/src/main/java/org/apache/metron/solr/dao/SolrSearchDao.java [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+            new SolrQuery().set("collection", collection));
+        documents.addAll(solrDocumentList.stream().map(this::toDocument).collect(Collectors.toList()));
+      }
+      return documents;
+    } catch (SolrServerException e) {
+      throw new IOException(e);
+    }
+  }
+
+  private SolrQuery buildSearchRequest(
```
[the remark this answers is not in the document]
Comment (merrimanr) [reply metron-platform/metron-solr/src/main/java/org/apache/metron/solr/dao/SolrSearchDao.java]: This was copied from ElasticsearchDao but I don't think Solr throws an exception in this case so I removed it.  Same applies to other similar comments.

File: metron-platform/metron-solr/src/main/java/org/apache/metron/solr/dao/SolrSearchDao.java [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+    if (facetFields.isPresent()) {
+      searchResponse.setFacetCounts(getFacetCounts(facetFields.get(), solrResponse));
+    }
+
+    if (LOG.isDebugEnabled()) {
+      String response;
+      try {
+        response = JSONUtils.INSTANCE.toJSON(searchResponse, false);
+      } catch (JsonProcessingException e) {
+        response = "???";
```
[the remark this answers is not in the document]
Comment (merrimanr) [reply metron-platform/metron-solr/src/main/java/org/apache/metron/solr/dao/SolrSearchDao.java]: There are several other places that use this pattern as well.  I changed it to `e.getMessage()` here but wanted you to be aware.

File: metron-platform/metron-solr/src/main/java/org/apache/metron/solr/dao/SolrSearchDao.java [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+    groupResponse.setGroupedBy(groupRequest.getGroups().get(0).getField());
+    groupResponse.setGroupResults(getGroupResults(groupRequest, 0, pivotFields));
+    return groupResponse;
+  }
+
+  private List<GroupResult> getGroupResults(GroupRequest groupRequest, int index, List<PivotField> pivotFields) {
+    List<Group> groups = groupRequest.getGroups();
+    List<GroupResult> searchResultGroups = new ArrayList<>();
+    final GroupOrder groupOrder = groups.get(index).getOrder();
+    Collections.sort(pivotFields, (o1, o2) -> {
```
[the remark this answers is not in the document]
Comment (merrimanr) [reply metron-platform/metron-solr/src/main/java/org/apache/metron/solr/dao/SolrSearchDao.java]: Done

File: metron-platform/metron-solr/src/main/java/org/apache/metron/solr/dao/SolrUpdateDao.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        this.client.add(solrUpdates);
+      }
+    } catch (SolrServerException e) {
+      throw new IOException(e);
+    }
+  }
+
+  private SolrInputDocument toSolrInputDocument(Document document) {
+    SolrInputDocument solrInputDocument = new SolrInputDocument();
+    document.getDocument().entrySet().forEach(entry ->
```
[the remark this answers is not in the document]
Comment (merrimanr) [reply metron-platform/metron-solr/src/main/java/org/apache/metron/solr/dao/SolrUpdateDao.java]: Done

File: metron-platform/metron-solr/src/test/java/org/apache/metron/solr/integration/SolrSearchIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+  protected InMemoryComponent startIndex() throws Exception {
+    solrComponent = new SolrComponent.Builder()
+        .addCollection("bro", "../metron-solr/src/test/resources/config/bro/conf")
+        .addCollection("snort", "../metron-solr/src/test/resources/config/snort/conf")
+        .build();
+    solrComponent.start();
+    return solrComponent;
+  }
+
+  @Override
```
[the remark this answers is not in the document]
Comment (merrimanr) [reply metron-platform/metron-solr/src/test/java/org/apache/metron/solr/integration/SolrSearchIntegrationTest.java]: Done

File: metron-platform/metron-solr/src/test/java/org/apache/metron/solr/integration/components/SolrComponent.java [hunk 11]
```diff
@@ -158,4 +162,16 @@ public boolean hasCollection(String collection) {
...
     return docs;
   }
+
+  public void addDocs(String collection, List<Map<String, Object>> docs)
+      throws IOException, SolrServerException {
+    CloudSolrClient solr = miniSolrCloudCluster.getSolrClient();
+    solr.setDefaultCollection(collection);
+    Collection<SolrInputDocument> solrInputDocuments = docs.stream().map(doc -> {
+      SolrInputDocument solrInputDocument = new SolrInputDocument();
+      doc.entrySet().forEach(entry -> solrInputDocument.addField(entry.getKey(), entry.getValue()));
```
[the remark this answers is not in the document]
Comment (merrimanr) [reply metron-platform/metron-solr/src/test/java/org/apache/metron/solr/integration/components/SolrComponent.java]: Done

**Comment by merrimanr** [comment]
I believe all feedback has been addressed.  The ElasticsearchDao refactor was not trivial and was more complex than the SolrDao.  There were more dependencies between the various Daos so it's not as simple as I would like.  Let me know what you think.

**Comment by justinleet** [comment]
I looked over the changes and refactor.  Thanks a lot for taking care of that, it looks a lot better and is definitely easier to grok, imo, than the old way.

I'm +1 on merging this into the feature branch.

**Comment by cestella** [comment]
well, I botched the stupid commit message and put `apache/metron` when I meant `apache/feature/METRON-1416-upgrade-solr`.  Can you close this PR @merrimanr ?

**Comment by cestella** [comment]
actually, it may be that feature branch PRs don't automatically close..or hmm, not sure.  Anyway, close it please :)

**Pull request closed without merging**